subpackages/property/source/unit_threaded/property.d

```diff
@@ -39,56 +38,67 @@ void check(alias F, int numFuncCalls = 100)
     scope gen = RndValueGen!(Parameters!F)(&random);
 
     auto input(Flag!"shrink" shrink = Yes.shrink) {
+
         scope string[] ret;
+
         static if(Parameters!F.length == 1 && canShrink!(Parameters!F[0])) {
             auto val = gen.values[0].value;
             auto shrunk = shrink ? val.shrink!F : val;
             ret ~= shrunk.text.idup;
             static if(isSomeString!(Parameters!F[0]))
                 ret[$-1] = `"` ~ ret[$-1] ~ `"`;
-        } else
+        } else {
             foreach(ref valueGen; gen.values) {
                 ret ~= valueGen.text;
             }
+        }
+
         return ret.join(", ");
     }
 
     foreach(i; 0 .. numFuncCalls) {
         bool pass;
 
-        () @trusted {  // catch throwable
-            try {
-                static if(isSafe!({ gen.genValues; }))
-                    () @safe { gen.genValues; }();
-                else
-                    gen.genValues;
-            } catch(Throwable t) {
-                throw new PropertyException("Error generating values\n" ~ t.toString, file, line, t);
-            }
-        }();
-
-        () @trusted { // catch throwable
-            try {
-                static if(isSafe!F)
-                    pass = () @safe { return F(gen.values); }();
-                else
-                    pass = F(gen.values);
-            } catch(Throwable t) {
-                // trying to shrink when an exeption is thrown is too much of a bother code-wise
+        safelyCatchThrowable!(
+            { gen.genValues; },
+            (t) @trusted { throw new PropertyException("Error generating values\n" ~ t.toString, file, line, t); }
+        );
+
+        pass = safelyCatchThrowable!(
+            () => F(gen.values),
+            (t) @trusted {
                 throw new UnitTestException(
                     text("Property threw. Seed: ", seed, ". Input: ", input(No.shrink), ". Message: ", t.msg),
                     file,
                     line,
                     t,
                     );
             }
-        }();
+        );
 
         if(!pass)
             throw new UnitTestException(text("Property failed. Seed: ", seed, ". Input: ", input), file, line);
     }
 }
 
+private auto safelyCatchThrowable(alias Func, alias Handler, A...)(auto ref A args) {
+    import std.traits: isSafe, ReturnType;
+
+    ReturnType!Func impl() {
+        try
+            return Func(args);
+        catch(Throwable t)
+            Handler(t);
+        assert(0);
+    }
+
+    static if(isSafe!Func)
+        return () @trusted { return impl; }();
+    else
+        return impl;
+}
```

Comment:
Oooh, this is cool.

However, you are still trusting `Handler` if `Func` is safe. Probably change static if to check both for safety.

Reply:
Huh, I literally just thought of that myself.

Reply:
looks good now.